state-chain/chains/src/sol/instruction_builder.rs

```diff
@@ -0,0 +1,313 @@
+			ComputeBudgetInstruction::set_compute_unit_price(compute_price),
+			ComputeBudgetInstruction::set_compute_unit_limit(compute_limit),
+		];
+
+		final_instructions.append(&mut instructions);
+
+		final_instructions
+	}
+
+	/// Create an instruction set to fetch from each `deposit_channel` being passed in.
+	/// Used to batch fetch from multiple deposit channels in a single transaction.
+	pub fn fetch_from(
+		deposit_channels: Vec<DepositChannel<Solana>>,
+		vault_program_data_account: SolAddress,
+		system_program_id: SolAddress,
+		agg_key: SolAddress,
+		nonce_account: SolAddress,
+		compute_price: SolAmount,
+	) -> Vec<SolInstruction> {
+		let instructions = deposit_channels
+			.into_iter()
+			.map(|deposit_channel| match deposit_channel.asset {
+				SolAsset::Sol => ProgramInstruction::get_instruction(
+					&VaultProgram::FetchNative {
+						seed: deposit_channel.channel_id.to_le_bytes().to_vec(),
+						bump: deposit_channel.state,
+					},
+					vec![
+						SolAccountMeta::new_readonly(vault_program_data_account.into(), false),
+						SolAccountMeta::new(agg_key.into(), true),
+						SolAccountMeta::new(deposit_channel.address.into(), false),
+						SolAccountMeta::new_readonly(system_program_id.into(), false),
+					],
+				),
+			})
+			.collect::<Vec<_>>();
+
+		Self::finalize(instructions, nonce_account.into(), agg_key.into(), compute_price)
+	}
+
+	/// Create an instruction set to `transfer` from our Vault account to a target account.
+	pub fn transfer(
+		to: TransferAssetParams<Solana>,
+		agg_key: SolAddress,
+		nonce_account: SolAddress,
+		compute_price: SolAmount,
+	) -> Vec<SolInstruction> {
+		let instructions = match to.asset {
+			SolAsset::Sol =>
+				vec![SystemProgramInstruction::transfer(&agg_key.into(), &to.to.into(), to.amount)],
+		};
+
+		Self::finalize(instructions, nonce_account.into(), agg_key.into(), compute_price)
+	}
+
+	/// Create an instruction set to rotate the current Vault agg key to the next key.
+	pub fn rotate_agg_key(
+		new_agg_key: SolAddress,
+		all_nonce_accounts: Vec<SolAddress>,
+		vault_program_data_account: SolAddress,
+		system_program_id: SolAddress,
+		upgrade_manager_program_data_account: SolAddress,
+		agg_key: SolAddress,
+		nonce_account: SolAddress,
+		compute_price: SolAmount,
+	) -> Vec<SolInstruction> {
+		let mut instructions = vec![
+			ProgramInstruction::get_instruction(
+				&VaultProgram::RotateAggKey { skip_transfer_funds: false },
+				vec![
+					SolAccountMeta::new(vault_program_data_account.into(), false),
+					SolAccountMeta::new(agg_key.into(), true),
+					SolAccountMeta::new(new_agg_key.into(), false),
+					SolAccountMeta::new_readonly(system_program_id.into(), false),
+				],
+			),
+			set_upgrade_authority(
+				upgrade_manager_program_data_account.into(),
+				&agg_key.into(),
+				Some(&new_agg_key.into()),
+			),
+		];
+		instructions.extend(all_nonce_accounts.into_iter().map(|nonce_account| {
+			SystemProgramInstruction::nonce_authorize(
```

Comment:
We call nonce_authorize on all nonce accounts here, and then in `finalize`, the `nonce_authorize` instruction is added again. Is this correct? (I guess 'all_nonce_accounts' means 'all except the nonce that is used to sign'?)

Reply:
Finalize calls `advance_nonce_account`, not `nonce_authorize`.